src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -1360,6 +1425,433 @@ private Object writeReplace() {
             return new CollSer(CollSer.IMM_MAP, array);
         }
     }
+
+    // Stable collections
+
+    @jdk.internal.ValueBased
+    static final class StableList<V>
+            extends AbstractImmutableList<StableValue<V>>
+            implements List<StableValue<V>> {
+
+        @Stable
+        private int size;
+        @Stable
+        private final V[] elements;
+        @Stable
+        private final byte[] sets;
+        private final Object[] mutexes;
+
+        @SuppressWarnings("unchecked")
+        private StableList(int size) {
+            assert size > 0;
+            this.elements = (V[]) new Object[size];
+            this.size = size;
+            this.sets = new byte[size];
+            this.mutexes = new Object[size];
+        }
+
+        @Override
+        public StableValue<V> get(int index) {
+            Objects.checkIndex(index, size);
+            return new StableValueElement<>(elements, sets, mutexes, index);
+        }
+
+        @Override
+        public int size() {
+            return size;
+        }
+
+        @Override
+        public boolean contains(Object o) {
+            Objects.requireNonNull(o);
+            return false;
+        }
+
+        @Override
+        public boolean containsAll(Collection<?> c) {
+            Objects.requireNonNull(c);
+            return c.isEmpty();
+        }
+
+        @Override
+        public int indexOf(Object o) {
+            Objects.requireNonNull(o);
+            return -1;
+        }
+
+        @Override
+        public int lastIndexOf(Object o) {
+            Objects.requireNonNull(o);
+            return -1;
+        }
+
+        V computeIfUnset(int index, IntFunction<? extends V> mapper) {
+            StableValueElement<V> element = new StableValueElement<>(elements, sets, mutexes, index);
+            return element.computeIfUnset(mapper);
+        }
+
+        static <V> List<StableValue<V>> create(int size) {
+            return new StableList<>(size);
+        }
+
+    }
+
+    // Internal interface used to indicate the presence of
+    // the computeIfUnset method that is unique to StableMap and StableEnumMap
+    interface HasComputeIfUnset<K, V> {
+        V computeIfUnset(K key, Function<? super K, ? extends V> mapper);
+    }
+
+    static final class StableMap<K, V>
```

Comment:
Question: can stable maps be implemented in terms of stable lists? After all, you need a stable backing array - and then you need to have a way to go from a key to an index in the stable array. The logic that does key->index conversion belongs to Map, but after that we should be able to just "delegate" to StableList?

Reply:
This is partially done but we could pull more on this string and unify the implementations.

Incidentally, it is also possible to unify the two implementation classes of `StableValue` so it becomes monomorphic.